x/fee_grant/internal/keeper/keeper.go

```diff
@@ -0,0 +1,121 @@
+func (k Keeper) GrantFeeAllowance(ctx sdk.Context, grant types.FeeAllowanceGrant) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FeeAllowanceKey(grant.Granter, grant.Grantee)
+	bz := k.cdc.MustMarshalBinaryBare(grant)
+	store.Set(key, bz)
+}
+
+// RevokeFeeAllowance removes an existing grant
+func (k Keeper) RevokeFeeAllowance(ctx sdk.Context, granter, grantee sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FeeAllowanceKey(granter, grantee)
+	store.Delete(key)
+}
+
+// GetFeeAllowance returns the allowance between the granter and grantee.
+// If there is none, it returns nil, nil.
+// Returns an error on parsing issues
+func (k Keeper) GetFeeAllowance(ctx sdk.Context, granter, grantee sdk.AccAddress) exported.FeeAllowance {
+	grant, found := k.GetFeeGrant(ctx, granter, grantee)
+	if !found {
+		return nil
+	}
+	return grant.Allowance
+}
+
+// GetFeeGrant returns entire grant between both accounts
+func (k Keeper) GetFeeGrant(ctx sdk.Context, granter sdk.AccAddress, grantee sdk.AccAddress) (types.FeeAllowanceGrant, bool) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FeeAllowanceKey(granter, grantee)
+	bz := store.Get(key)
+	if len(bz) == 0 {
+		return types.FeeAllowanceGrant{}, false
+	}
+
+	var grant types.FeeAllowanceGrant
+	k.cdc.MustUnmarshalBinaryBare(bz, &grant)
+	return grant, true
+}
+
+// IterateAllGranteeFeeAllowances iterates over all the grants from anyone to the given grantee.
+// Callback to get all data, returns true to stop, false to keep reading
+func (k Keeper) IterateAllGranteeFeeAllowances(ctx sdk.Context, grantee sdk.AccAddress, cb func(types.FeeAllowanceGrant) bool) error {
+	store := ctx.KVStore(k.storeKey)
+	prefix := types.FeeAllowancePrefixByGrantee(grantee)
+	iter := sdk.KVStorePrefixIterator(store, prefix)
+	defer iter.Close()
+
+	stop := false
+	for ; iter.Valid() && !stop; iter.Next() {
+		bz := iter.Value()
+		var grant types.FeeAllowanceGrant
+		err := k.cdc.UnmarshalBinaryBare(bz, &grant)
+		if err != nil {
+			return err
+		}
+		stop = cb(grant)
+	}
+	return nil
+}
+
+// IterateAllFeeAllowances iterates over all the grants in the store.
+// Callback to get all data, returns true to stop, false to keep reading
+// Calling this without pagination is very expensive and only designed for export genesis
+func (k Keeper) IterateAllFeeAllowances(ctx sdk.Context, cb func(types.FeeAllowanceGrant) bool) error {
+	store := ctx.KVStore(k.storeKey)
+	iter := sdk.KVStorePrefixIterator(store, types.FeeAllowanceKeyPrefix)
+	defer iter.Close()
+
+	stop := false
+	for ; iter.Valid() && !stop; iter.Next() {
+		bz := iter.Value()
+		var grant types.FeeAllowanceGrant
+		err := k.cdc.UnmarshalBinaryBare(bz, &grant)
+		if err != nil {
+			return err
+		}
+		stop = cb(grant)
+	}
+	return nil
+}
+
+// UseGrantedFees will try to pay the given fee from the granter's account as requested by the grantee
+func (k Keeper) UseGrantedFees(ctx sdk.Context, granter, grantee sdk.AccAddress, fee sdk.Coins) bool {
```

Comment:
question: why does this function return a `bool` instead of an `error`?

Reply:
The only real use was in https://github.com/cosmos/cosmos-sdk/blob/5ca80b8bf49b729cc1711a145f954aaf926a90be/x/fee_grant/internal/ante/fee.go#L68-L72 and I figured it was yes/no.

I can return an error instead to give more info